ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Aux.hs

```diff
@@ -0,0 +1,545 @@
+mkBodyState state = CC.BodyState {
+      CC.utxo            = CC.cvsUtxo            state
+    , CC.updateState     = CC.cvsUpdateState     state
+    , CC.delegationState = CC.cvsDelegationState state
+    }
+
+mkBodyEnv :: Gen.Config
+          -> Update.ProtocolParameters
+          -> CC.SlotNumber
+          -> CC.BodyEnvironment
+mkBodyEnv cfg params slotNo = CC.BodyEnvironment {
+      CC.protocolMagic      = reAnnotateMagic $ Gen.configProtocolMagic cfg
+    , CC.utxoConfiguration  = Gen.configUTxOConfiguration cfg
+    , CC.k                  = Gen.configK cfg
+    , CC.allowedDelegators  = allowedDelegators cfg
+    , CC.protocolParameters = params
+    , CC.currentEpoch       = CC.slotNumberEpoch
+                                (Gen.configEpochSlots cfg)
+                                slotNo
+    }
+
+applyEpochTransition :: Gen.Config
+                     -> CC.SlotNumber
+                     -> CC.ChainValidationState
+                     -> CC.ChainValidationState
+applyEpochTransition cfg slotNo state = state {
+      CC.cvsUpdateState = CC.epochTransition
+                            (mkEpochEnvironment cfg state)
+                            (CC.cvsUpdateState state)
+                            slotNo
+    }
+
+validateHeader :: MonadError CC.ChainValidationError m
+               => CC.ValidationMode
+               -> U.Iface.State -> CC.AHeader ByteString -> m ()
+validateHeader validationMode updState hdr =
+    flip runReaderT validationMode $
+      CC.headerIsValid updState hdr
+
+validateBody :: MonadError CC.ChainValidationError m
+             => CC.ValidationMode
+             -> CC.ABlock ByteString
+             -> CC.BodyEnvironment -> CC.BodyState -> m CC.BodyState
+validateBody validationMode block bodyEnv bodyState =
+    flip runReaderT validationMode $
+      CC.updateBody bodyEnv bodyState block
+
+validateBlock :: MonadError CC.ChainValidationError m
+              => Gen.Config
+              -> CC.ValidationMode
+              -> CC.ABlock ByteString
+              -> CC.HeaderHash
+              -> CC.ChainValidationState -> m CC.ChainValidationState
+validateBlock cfg validationMode block blkHash state = do
+    validateHeader validationMode updState (CC.blockHeader block)
+    bodyState' <- validateBody validationMode block bodyEnv bodyState
+    return state {
+          CC.cvsLastSlot        = CC.blockSlot block
+        , CC.cvsPreviousHash    = Right blkHash
+        , CC.cvsUtxo            = CC.utxo            bodyState'
+        , CC.cvsUpdateState     = CC.updateState     bodyState'
+        , CC.cvsDelegationState = CC.delegationState bodyState'
+        }
+  where
+    updState  = CC.cvsUpdateState state
+    bodyEnv   = mkBodyEnv cfg (getProtocolParams state) (CC.blockSlot block)
+    bodyState = mkBodyState state
+
+-- | Apply a boundary block
+--
+-- We don't validate boundary blocks and so this cannot fail.
+--
+-- TODO:
+-- * Should we check that the prev hash lines up?
```

Comment:
Comment: we are checking this elsewhere (in chain sync client).

Reply:
Modified this code to call `updateChainBoundary`, which _does_ check the hashes. Additionally, added this to `validateBLlck`:

```haskell
    -- TODO: How come this check isn't done in 'updateBlock'
    -- (but it /is/ done in 'updateChainBoundary')?
    --
    -- TODO: It could be argued that hash checking isn't part of consensus /or/
    -- the ledger. If we take that point of view serious, we should think about
    -- what that third thing is precisely and what it's responsibilities are.
    case ( CC.cvsPreviousHash state
         , unAnnotated $ CC.aHeaderPrevHash (CC.blockHeader block)
         ) of
      (Left gh, hh) ->
         throwError $ CC.ChainValidationExpectedGenesisHash gh hh
      (Right expected, actual) ->
         unless (expected == actual) $
           throwError $ CC.ChainValidationInvalidHash expected actual
```